include/internal/quic_fc.h

```diff
@@ -0,0 +1,266 @@
+ * connection-level flow control.
+ */
+uint64_t ossl_quic_txfc_get_credit_local(QUIC_TXFC *txfc);
+
+/*
+ * Consume num_bytes of credit. This is the 'On TX' operation. This should be
+ * called when we transmit any controlled bytes. Calling this with an argument
+ * of 0 is a no-op.
+ *
+ * We must never transmit more controlled bytes than we are in credit for (see
+ * the return value of ossl_quic_txfc_get_credit()). If you call this function
+ * with num_bytes greater than our current credit, this function consumes the
+ * remainder of the credit and returns 0. This indicates a serious programming
+ * error on the caller's part. Otherwise, the function returns 1.
+ *
+ * If called on a stream-level TXFC, ossl_quic_txfc_consume_credit() is called
+ * on the connection-level TXFC also. If the call to that function on the
+ * connection-level TXFC returns zero, this function will also return zero.
+ */
+int ossl_quic_txfc_consume_credit(QUIC_TXFC *txfc, uint64_t num_bytes);
+
+/*
+ * Like ossl_quic_txfc_consume_credit(), but when called on a stream-level TXFC,
+ * consumes only from the stream-level credit and does not inform the
+ * connection-level TXFC.
+ */
+int ossl_quic_txfc_consume_credit_local(QUIC_TXFC *txfc, uint64_t num_bytes);
+
+/*
+ * This flag is provided for convenience. A caller is not required to use it. It
+ * is a boolean flag set whenever our credit drops to zero. If clear is 1, the
+ * flag is cleared. The old value of the flag is returned. Callers may use this
+ * to determine if they need to send a DATA_BLOCKED or STREAM_DATA_BLOCKED
+ * frame, which should contain the value returned by ossl_quic_txfc_get_cwm().
+ */
+int ossl_quic_txfc_has_become_blocked(QUIC_TXFC *txfc, int clear);
+
+/*
+ * Get the current CWM value. This is mainly only needed when generating a
+ * DATA_BLOCKED or STREAM_DATA_BLOCKED frame, or for diagnostic purposes.
+ */
+uint64_t ossl_quic_txfc_get_cwm(QUIC_TXFC *txfc);
+
+/*
+ * Get the current spent watermark (SWM) value. This is purely for diagnostic
+ * use and should not be needed in normal circumstances.
+ */
+uint64_t ossl_quic_txfc_get_swm(QUIC_TXFC *txfc);
+
+/*
+ * RX Flow Controller (RXFC)
+ * =========================
+ */
+typedef struct quic_rxfc_st QUIC_RXFC;
+
+struct quic_rxfc_st {
+    /*
+     * swm is the sent/received watermark, which tracks how much we have
+     * received from the peer. rwm is the retired watermark, which tracks how
+     * much has been passed to the application. esrwm is the rwm value at which
+     * the current auto-tuning epoch started. hwm is the highest stream length
+     * (STREAM frame offset + payload length) we have seen from a STREAM frame
+     * yet.
+     */
+    uint64_t        cwm, swm, rwm, esrwm, hwm, cur_window_size, max_window_size;
+    OSSL_TIME       epoch_start;
+    OSSL_TIME     (*now)(void *arg);
+    void           *now_arg;
+    QUIC_RXFC      *parent;
```

Comment:
It seems not entirely OpenSSL style that we line up on the alphanumeric part of the variable names here.